### PR TITLE
#8778 Feature Grid filter by viewport

### PR DESCRIPTION
**Pull request by alex-fko**
## Description
This PR adds an additional filter type to the feature editor records: dynamic filter by viewport, updated on each map view change.

One important note about implementation: any other spatial filter applied to the request will be added as an "AND" condition.
E.g. when user selects viewport filter via advanced filters or draw rectangle on the map to select features. In all these cases condition will be `Drawn box AND Viewport rectangle must intersect feature` or `Advanced filter viewport AND Dynamic viewport must intersect feature`.

Tool is disabled by default and not available in toolbar. Please use following configuration to activate button and activate tool by default:
```
{
  "name": "FeatureEditor",
  "cfg": {
    "filterByViewport": true,
    "showFilterByViewportTool": true
  }
}
```

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#8778 

**What is the new behavior?**
New tool is available in toolbar once active in plugin configuration.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
Looks quite good.

- [ ] For the icon please look at this conversation https://github.com/geosolutions-it/MapStore2/pull/8788#issuecomment-1310072554 
- [ ] The filter should be reset together with the other tools (e.g. sync with map). Instead switching from one map to the other I see only it activated. Please find out how the other tools are reset and do the same.
- [x] Make the tool visible
- [x] I should change the text into "Filter data by the current viewport" - "Do not filter data by the current viewport" ("Activate viewport filter" is ambigous and messes up the translations --> they become something like "activate the filter of the view" )

**Comment by offtherailz** [comment]
Hi alexander, the icon has been added here  https://github.com/geosolutions-it/MapStore2/pull/8807 , please apply the change and request the review when you are ready
Thank you very much.

**Comment by alex-fko** [comment]
Thanks @offtherailz, I've applied new icon, all other points are addressed too.

**Review by offtherailz — CHANGES_REQUESTED** [review]
Good. We have only a couple of issues left: 
- [ ] The filter should be disabled in 3D mode
- [ ] The filter should be deactivated (reset) together with the other tools. I see it remains active all the time. This can cause
     - Issues when switching between different contexts (e.g. with this button not visible, but the tool that could stay active from a previous usage in a context where it was present and the user enabled)
     - Issues when switching from 3D to 2D

Please test the two workflow above to verify that the tool is consistent in any case

- [ ] Moreover there is another issue, more a UX problem, because we have in the results also the ones hidden by the table you drag and drop from the bottom. 
Looking at the current API of Openlayers and mapstore, it looks difficout to allow this. 
apparently calculateExtent method from view do not allow to pass a custom padding. The only way I can imagine we can fix it is to synchronize the padding with the current mapLayout state that holds the size in pixels of the tools around. 
Anyway this can easely become tricky and involve other tools that use padding (zoomToExtent/Indentify) and have to be discussed with the team. Do you have any other idea  @alex-fko ?

@tdipisa do you think we can live with this last Ux problem for the moment?

**Comment by tdipisa** [comment]
@offtherailz 

>@tdipisa do you think we can live with this last Ux problem for the moment?

behind certain extents you cannot expect I can provide an useful feedback on a topic simply mentioning me in the comment. That's something you should raise during our periodic sync maybe showing me what you are referring to :)

**Comment by alex-fko** [comment]
@offtherailz "The filter should be deactivated (reset) together with the other tools", I see your point. Each location change should reset value back to the default one defined in plugin configuration.

Regarding bbox calculation and the fact that filter includes part of viewport covered by the panel itself: I think it should be doable by utilizing  GET_COORDINATES_FROM_PIXEL_HOOK: Get coordinates of the ponts for the rectangle not covered with panel. Convert them into bbox and use it in request.

**Comment by offtherailz** [comment]
Using `GET_COORDINATES_FROM_PIXEL_HOOK` may be hard and complicate a lot the things.
Moreover I see from this issue that the way I suggested (by setting the padding in the map) has been fixed and properly works only for OL > 6.6.0  (we are on ol 5.3.0 now).

**Review by offtherailz — CHANGES_REQUESTED** [review]
Looks almost perfect. I noticed this glitch here (filter button that alternate blue and green color): 

https://user-images.githubusercontent.com/1279510/201884874-205036b9-818e-4f80-9659-5e3ed63d2637.mp4

Do you think we can fix it?

**Comment by alex-fko** [comment]
@offtherailz it looks like it's intended, it behaves same way on production now: once there is any filter (not necessarily via advanced search) - this button turns green. In case if there is no advanced filter or, spatial via drawn shape on the map, or simple via inputs above properties - it gets back to inactive blue state.

**Comment by alex-fko** [comment]
@offtherailz thanks, it should not flick now

File: web/client/utils/FilterUtils.js [hunk 1]
```diff
@@ -1119,6 +1119,7 @@ export const isFilterValid = (f = {}) =>
     || (f.simpleFilterFields && f.simpleFilterFields.length > 0)
     || (f.spatialField && f.spatialField.geometry && f.spatialField.operation ||
         isArray(f.spatialField) && findIndex(f.spatialField, field => field.operation && field.geometry) > -1)
+    || (f.viewportFilter && f.viewportFilter.geometry && f.viewportFilter.operation)
```
Comment (offtherailz) [inline web/client/utils/FilterUtils.js:1122]: Sorry, is this needed anymore?

**Review by offtherailz — CHANGES_REQUESTED** [review]
Last change: 
- [ ] change mapView !== 'cesium' into a selector called `isFilterByViewportSupported`. Makes everything more explicit and resistent to future evolutions.

**Review by offtherailz — APPROVED** [review]

**Comment by offtherailz** [comment]
@ElenaGallo, could you please test this on  DEV ? Thank you

**Comment by ElenaGallo** [comment]
Test passed on DEV

**Pull request merged**